fcrepo-clients/src/main/java/edu/unc/lib/dl/util/DerivativeService.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2008 The University of North Carolina at Chapel Hill
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.util;
+
+import static edu.unc.lib.dl.fcrepo4.RepositoryPathConstants.HASHED_PATH_DEPTH;
+import static edu.unc.lib.dl.fcrepo4.RepositoryPathConstants.HASHED_PATH_SIZE;
+import static edu.unc.lib.dl.fcrepo4.RepositoryPaths.idToPath;
+import static edu.unc.lib.dl.model.DatastreamType.getByIdentifier;
+import static edu.unc.lib.dl.model.StoragePolicy.EXTERNAL;
+import static org.springframework.util.Assert.notNull;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import edu.unc.lib.dl.fedora.PID;
+import edu.unc.lib.dl.model.DatastreamType;
+
+/**
+ * Service which locates and returns file information about derivatives for
+ * repository objects.
+ *
+ * @author bbpennel
+ *
+ */
+public class DerivativeService {
+
+    private String derivativeDir;
+
+    private volatile static List<DatastreamType> derivativeTypes;
```

Comment:
not challenging your choice here, but just curious about why the derivativeTypes need to be stored in main memory

Reply:
It is going to be a highly accessed value, since every request for a thumbnail from the UI and every datastream at indexing time will get compared against it, so figured it was worth storing.